pennylane/interfaces/batch/jax.py

```diff
@@ -0,0 +1,257 @@
+                dy = args[-1]
+
+                for t, a in zip(tapes, p):
+                    new_tapes.append(t.copy(copy_operations=True))
+                    new_tapes[-1].set_parameters(a)
+                    new_tapes[-1].trainable_params = t.trainable_params
+
+                vjp_tapes, processing_fn = qml.gradients.batch_vjp(
+                    new_tapes,
+                    dy,
+                    gradient_fn,
+                    reduction="append",
+                    gradient_kwargs=gradient_kwargs,
+                )
+
+                partial_res = execute_fn(vjp_tapes)[0]
+                res = processing_fn(partial_res)
+                return np.concatenate(res)
+
+            args = tuple(params) + (g,)
+            vjps = host_callback.call(
+                non_diff_wrapper,
+                args,
+                result_shape=jax.ShapeDtypeStruct((total_params,), dtype),
+            )
+
+            param_idx = 0
+            res = []
+
+            # Group the vjps based on the parameters of the tapes
+            for p in params:
+                res.append(vjps[param_idx : param_idx + len(p)])
+                param_idx += len(p)
+
+                # Unpack if we have the vjp for a single parameter
+                if len(p) == 1:
+                    res[-1] = res[-1][0]
+
+            # Unwrap partial results into arrays if need be
+            need_unwrapping = any(r.ndim != 0 for r in res)
+            if need_unwrapping:
+                unwrapped_res = []
+                for r in res:
+                    if r.ndim != 0:
+                        r = [jnp.array(p) for p in r]
+                    unwrapped_res.append(r)
+
+                res = unwrapped_res
+
+            return (tuple(res),)
+
+        # Gradient function is a device method.
+        with qml.tape.Unwrap(*tapes):
+            jacs = gradient_fn(tapes, **gradient_kwargs)
+
+        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
+        res = [[jnp.array(p) for p in v] for v in vjps]
+        return (tuple(res),)
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    return wrapped_exec(params)
+
+
+# The execute function in forward mode
+def _execute_with_fwd(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+
+    # Only have scalar outputs
+    total_size = len(tapes)
+
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        def wrapper(p):
+            """Compute the forward pass by returning the jacobian too."""
+            new_tapes = []
+
+            for t, a in zip(tapes, p):
+                new_tapes.append(t.copy(copy_operations=True))
+                new_tapes[-1].set_parameters(a)
+
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+            # On the forward execution return the jacobian too
+            return np.stack(res), jacs

```

Comment:
oh wait, the wrapper is allowed to return a tuple??

Maybe this is a path forwards for supporting non-scalar return types :)

Reply:
Yep, this should definitely be the way. :) It will likely be good to have a convenient way to determine the output shape of quantum functions to support non-scalar return types but should be doable.